File: fp/anemic/src/main/clojure/de/hhu/anemic/warenkorb/domain/model/warenkorb/Warenkorb.clj
(ns de.hhu.anemic.warenkorb.domain.model.warenkorb.Warenkorb
  (:require [de.hhu.anemic.warenkorb.domain.model.anzahl.Anzahl :as anzahl]
            [de.hhu.anemic.warenkorb.domain.model.preis.Preis :as preis]
            [de.hhu.anemic.warenkorb.domain.model.warenkorb.WarenkorbID :as warenkorb-id]
            [de.hhu.anemic.warenkorb.domain.model.warenkorb.warenkorbzeile.Warenkorbzeile :as warenkorbzeile])
  (:import (de.hhu.anemic.warenkorb.domain.model.kunde.KundeID KundeID)))

(defrecord Warenkorb [id kunde-id warenkorbzeilen gesamt-preis max-einkaufswert])

(defn validiere-warenkorb [warenkorb]
  (let [gesamt-preis (:gesamt-preis warenkorb)
        max-einkaufswert (:max-einkaufswert warenkorb)]
    (when (or (nil? (:id warenkorb))
              (nil? (:kunde-id warenkorb))
              (nil? gesamt-preis)
              (nil? max-einkaufswert)
              (preis/ist-groesser-als? gesamt-preis max-einkaufswert))
      (throw (IllegalArgumentException. "Fehlerhafter Warenkorb")))))

(defn erstelle-neuen-warenkorb [kunde-id max-einkaufswert]
  (let [neuer-warenkorb (->Warenkorb (warenkorb-id/erstelleWarenkorbID)
                                     (KundeID kunde-id)
                                     []
                                     (preis/erstelle-preis "0.00" "EUR")
                                     (preis/erstelle-preis max-einkaufswert "EUR"))]
    (validiere-warenkorb neuer-warenkorb)
    neuer-warenkorb))

(defn erstelle-warenkorb [warenkorb neue-warenkorbzeilen neuer-gesamtpreis]
  (->Warenkorb (:id warenkorb)
               (:kunde-id warenkorb)
               neue-warenkorbzeilen
               neuer-gesamtpreis
               (:max-einkaufswert warenkorb)))

(defn finde-zeile-zu-artikel [warenkorb artikel-id]
  (first (filter #(= artikel-id (:artikel-id %)) (:warenkorbzeilen warenkorb))))

(defn fuege-warenkorbzeile-hinzu [warenkorb artikel anzahl]
  (let [neue-warenkorbzeilen (conj (:warenkorbzeilen warenkorb)
                                   (warenkorbzeile/erstelle-warenkorbzeile (:id artikel)
                                                                           (anzahl/erstelle-anzahl anzahl)
                                                                           (:preis artikel)
                                                                           (anzahl/erstelle-anzahl 10)))
        neuer-gesamtpreis (preis/erhoehe-um (:gesamt-preis warenkorb)
                                            (warenkorbzeile/berechne-gesamtpreis (last neue-warenkorbzeilen)))]
    (erstelle-warenkorb warenkorb neue-warenkorbzeilen neuer-gesamtpreis)))

(defn fuege-hinzu [warenkorb artikel anzahl]
  (let [zeile-mit-artikel (finde-zeile-zu-artikel warenkorb (:id artikel))
        gesamtpreis-von-artikel (if zeile-mit-artikel
                                  (warenkorbzeile/berechne-gesamtpreis zeile-mit-artikel)
                                  (preis/erstelle-preis (:preis artikel) "EUR"))
        neuer-warenkorb (if zeile-mit-artikel
                          (conj (remove #(= artikel (:id %)) (:warenkorbzeilen warenkorb))
                                (warenkorbzeile/erhoehe-um zeile-mit-artikel anzahl))
                          (fuege-warenkorbzeile-hinzu warenkorb artikel anzahl))]
    (erstelle-warenkorb warenkorb (:warenkorbzeilen neuer-warenkorb) gesamtpreis-von-artikel)))

(defn reduziere [warenkorb artikel anzahl]
  (let [zeile-mit-artikel (finde-zeile-zu-artikel warenkorb (:id artikel))]
    (when zeile-mit-artikel
      (let [neue-warenkorbzeile (warenkorbzeile/reduziere-um zeile-mit-artikel anzahl)]
        (when (= 0 (:anzahl neue-warenkorbzeile))
          (conj (remove #(= artikel (:id %)) (:warenkorbzeilen warenkorb)) neue-warenkorbzeile))
        (erstelle-warenkorb warenkorb (:warenkorbzeilen warenkorb) (warenkorbzeile/berechne-gesamtpreis zeile-mit-artikel))))))

(defn entferne [warenkorb artikel]
  (let [zeile-mit-artikel (finde-zeile-zu-artikel warenkorb (:id artikel))]
    (when zeile-mit-artikel
      (let [gesamtpreis-von-artikel (warenkorbzeile/berechne-gesamtpreis zeile-mit-artikel)]
        (erstelle-warenkorb warenkorb (vec (remove #(= artikel (:id %)) (:warenkorbzeilen warenkorb)))
                                  (preis/reduziere-um (:gesamt-preis warenkorb) gesamtpreis-von-artikel))))))

File: oop/rich/src/main/java/de/hhu/rich/warenkorb/domain/model/warenkorb/Warenkorb.java
package de.hhu.rich.warenkorb.domain.model.warenkorb;

import de.hhu.rich.warenkorb.domain.model.PflichtfeldFehlt;
import de.hhu.rich.warenkorb.domain.model.artikel.Artikel;
import de.hhu.rich.warenkorb.domain.model.artikel.ArtikelID;
import de.hhu.rich.warenkorb.domain.model.common.anzahl.Anzahl;
import de.hhu.rich.warenkorb.domain.model.common.preis.Preis;
import de.hhu.rich.warenkorb.domain.model.kunde.KundeID;
import de.hhu.rich.warenkorb.domain.model.warenkorb.warenkorbzeile.Warenkorbzeile;
import de.hhu.rich.warenkorb.domain.model.warenkorb.warenkorbzeile.WarenkorbzeileID;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

public class Warenkorb {

    private final WarenkorbID ID;
    private final KundeID kundeID;
    private final List<Warenkorbzeile> warenkorbzeilen;
    private Preis gesamtPreis;
    private final Preis maxEinkaufswert;

    public Warenkorb(WarenkorbID ID, KundeID kundeID, Preis maxEinkaufswert) {
        this.ID = ID;
        this.kundeID = kundeID;
        this.warenkorbzeilen = new ArrayList<>();
        this.gesamtPreis = new Preis(new BigDecimal(0), "EUR");
        this.maxEinkaufswert = maxEinkaufswert;
        this.validiere();
    }

    public Warenkorb(WarenkorbID ID,
                     KundeID kundeID,
                     List<Warenkorbzeile> warenkorbzeilen,
                     Preis gesamtPreis,
                     Preis maxEinkaufswert) {
        this.ID = ID;
        this.kundeID = kundeID;
        this.warenkorbzeilen = new ArrayList<>(warenkorbzeilen);
        this.gesamtPreis = gesamtPreis;
        this.maxEinkaufswert = maxEinkaufswert;
        this.validiere();
    }

    public WarenkorbID getWarenkorbID() {
        return ID;
    }

    public KundeID getKundeID() {
        return kundeID;
    }

    public List<Warenkorbzeile> getWarenkorbzeilen() {
        return warenkorbzeilen;
    }

    public Preis getGesamtPreis() {
        return gesamtPreis;
    }

    public Preis getMaxEinkaufswert() {
        return maxEinkaufswert;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Warenkorb warenkorb = (Warenkorb) o;
        return ID.equals(warenkorb.ID);
    }

    @Override
    public int hashCode() {
        if(ID != null) return ID.hashCode();
        else return super.hashCode();
    }

    @Override
    public String toString() {
        return "Warenkorb{" +
                "id=" + ID +
                ", kundeID=" + kundeID +
                ", warenkorbzeilen=" + warenkorbzeilen +
                ", gesamtPreis=" + gesamtPreis +
                ", maxEinkaufswert=" + maxEinkaufswert +
                '}';
    }

    public Warenkorbzeile findeZeileZu(ArtikelID artikelId) {
        return warenkorbzeilen.stream()
                .filter(a -> a.getArtikelID().equals(artikelId))
                .findFirst()
                .orElse(null);
    }

    public void fuegeHinzu(Artikel artikel, Anzahl anzahl) {
        Warenkorbzeile zeileMitArtikel = findeZeileZu(artikel.getID());

        if(zeileMitArtikel != null) {
            zeileMitArtikel.erhoeheUm(anzahl);
        } else {
            warenkorbzeilen.add(
                    new Warenkorbzeile(
                            new WarenkorbzeileID(UUID.randomUUID()),
                            artikel.getID(),
                            anzahl,
                            artikel.getPreis(),
                            new Anzahl(10)));
        }
        Preis gesamtpreisVonArtikel = findeZeileZu(artikel.getID()).berechneGesamtpreis();
        this.gesamtPreis = gesamtPreis.erhoeheUm(gesamtpreisVonArtikel);
        this.validiere();
    }

    public void reduziere(Artikel artikel, Anzahl anzahl) {
        Warenkorbzeile zeileMitArtikel = findeZeileZu(artikel.getID());

        if (zeileMitArtikel != null) {
            zeileMitArtikel.reduziereUm(anzahl);
            if (zeileMitArtikel.getAnzahl() == null) {
                warenkorbzeilen.remove(zeileMitArtikel);
            }
            Preis gesamtpreisVonArtikel = zeileMitArtikel.berechneGesamtpreis();
            this.gesamtPreis = gesamtPreis.reduziereUm(gesamtpreisVonArtikel);
        }
        validiere();
    }

    public void entferne(Artikel artikel) {
        Warenkorbzeile zeileMitArtikel = findeZeileZu(artikel.getID());

        if (zeileMitArtikel != null) {
            Preis gesamtpreisVonArtikel = zeileMitArtikel.berechneGesamtpreis();
            warenkorbzeilen.remove(zeileMitArtikel);
            this.gesamtPreis = gesamtPreis.reduziereUm(gesamtpreisVonArtikel);
        }
        validiere();
    }

    private void validiere() {
        if (ID == null) {
            throw new PflichtfeldFehlt("WarenkorbzeileID darf nicht null sein!");
        }
        if (kundeID == null) {
            throw new PflichtfeldFehlt("KundeId darf nicht null sein!");
        }
        if (gesamtPreis == null) {
            throw new PflichtfeldFehlt("Gesamt darf nicht null sein!");
        }
        if (maxEinkaufswert == null) {
            throw new PflichtfeldFehlt("Maximaler Einkaufswert darf nicht null sein!");
        }
        if (getGesamtPreis().istGroesserAls(maxEinkaufswert)) {
            throw new IllegalStateException("Maximaler Einkaufswert ist überschritten.");
        }
    }
}

